File: app/src/main/java/com/friendlypos/application/bluetooth/PrinterService.java
package com.friendlypos.application.bluetooth;


import android.app.Service;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Binder;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.util.Log;

import com.friendlypos.application.FriendlyApp;
import com.friendlypos.principal.activity.MenuPrincipal;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Set;
import java.util.UUID;
import java.util.Vector;


public class PrinterService extends Service {

    public static final String CLASS_NAME = "com.friendlypos.application.bluetooth.PrinterService";
    public static final String BROADCAST_CLASS = "com.friendlypos.printbill";
    private static final UUID UUID_SPP = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB");

    public static boolean stop(MenuPrincipal context){
        if (context.isServiceRunning(CLASS_NAME)) {
            Intent serviceIntent = new Intent(context, PrinterService.class);
            context.stopService(serviceIntent);
            return true;
        }
        return false;
    }



    private static final String TAG = "PrintService";

    private BluetoothAdapter mBluetoothAdapter;
    public static final String BT_DEVICE = "btdevice";
    public static final int STATE_NONE = 0; // we're doing nothing
    public static final int STATE_LISTEN = 1; // now listening for incoming
    // connections
    public static final int STATE_CONNECTING = 2; // now initiating an outgoing
    // connection
    public static final int STATE_CONNECTED = 3; // now connected to a remote
    // device
    private ConnectThread mConnectThread;
    private static ConnectedThread mConnectedThread;

    public static Handler mHandler = null;
    public static Handler mHandler2 = null;
    public static int mState = STATE_NONE;
    public static String deviceName;
    public static BluetoothDevice device = null;

    private String mMacAddress;

    private Context mContext;
    private BluetoothSocket onSocket;
    public static void startRDService(Context context, String prefExternalDisplay){
        if (prefExternalDisplay == null || prefExternalDisplay.equals("")) return;

        Intent serviceIntent = new Intent(context, PrinterService.class);
     serviceIntent.putExtra(BT_DEVICE, prefExternalDisplay);
        context.startService(serviceIntent);
    }

    @Override
    public void onCreate() {
        Log.i(TAG, "Service started");
        this.registerReceiver(onBroadcast, new IntentFilter(PrinterService.BROADCAST_CLASS));

        mContext = this;
        mHandler2 = new Handler();

        super.onCreate();
    }


    @Override
    public IBinder onBind(Intent intent) {
        Log.d(TAG, "ONBIND");
        return mBinder;
    }

    public class LocalBinder extends Binder {
        PrinterService getService() {
            return PrinterService.this;
        }
    }



    private final IBinder mBinder = new LocalBinder();


    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.d(TAG, "Onstart Command");
        mBluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
        String chosenDevice = intent.getStringExtra(BT_DEVICE);
        if (mBluetoothAdapter != null && !chosenDevice.equals("")) {
            Set<BluetoothDevice> pairedDevices = mBluetoothAdapter.getBondedDevices();
            if (pairedDevices.size() > 0) {
                for (BluetoothDevice d : pairedDevices) {
                    if (d.getAddress().equals(chosenDevice))
                        device = d;
                }
            }
            if (device == null){
                Log.d(TAG, "No device... stopping");
                return START_STICKY_COMPATIBILITY;
            }
            deviceName = device.getName();
            mMacAddress = device.getAddress();
            if (mMacAddress != null && mMacAddress.length() > 0) {
                Log.d(TAG, "Connecting to: "+deviceName);
                connectToDevice(mMacAddress);
            } else {
                Log.d(TAG, "No macAddress... stopping");
                stopSelf();
                return START_STICKY_COMPATIBILITY;
            }
        }

        return START_STICKY;
    }

    private synchronized void connectToDevice(String macAddress) {
        Log.d(TAG, "Connecting... ");
        BluetoothDevice device = mBluetoothAdapter.getRemoteDevice(macAddress);
        if (mState == STATE_CONNECTING) {
            if (mConnectThread != null) {
                mConnectThread.cancel();
                mConnectThread = null;
            }
        }

        // Cancel any thread currently running a connection
        if (mConnectedThread != null) {
            mConnectedThread.cancel();
            mConnectedThread = null;
        }
        mConnectThread = new ConnectThread(device);
        mConnectThread.start();
        setState(STATE_CONNECTING);
    }

    private void setState(int state) {
        PrinterService.mState = state;
    }

    public synchronized void stop() {
        setState(STATE_NONE);
        mHandler2.removeCallbacks(reconnect);

        if (mConnectThread != null) {
            mConnectThread.cancel();
            mConnectThread = null;
        }

        if (mConnectedThread != null) {
            mConnectedThread.cancel();
            mConnectedThread = null;
        }
        if (mBluetoothAdapter != null) {
            mBluetoothAdapter.cancelDiscovery();
        }
        stopSelf();
    }

    @Override
    public boolean stopService(Intent name) {

        setState(STATE_NONE);
        if (mConnectThread != null) {
            mConnectThread.cancel();
            mConnectThread = null;
        }

        if (mConnectedThread != null) {
            mConnectedThread.cancel();
            mConnectedThread = null;
        }
        mBluetoothAdapter.cancelDiscovery();
        return super.stopService(name);
    }

    private void connectionFailed() {
        Log.d(TAG, "Connection Failed");
        //RaceResultsDisplayService.this.stop();
        // Post to UI that connection is off
        if (mState == STATE_NONE) return;
        reconnect();
    }

    public void connectionLost() {
        Log.d(TAG, "Connection Lost");
        if (mConnectThread != null) {
            mConnectThread.cancel();
            mConnectThread = null;
        }

        if (mConnectedThread != null) {
            mConnectedThread.cancel();
            mConnectedThread = null;
        }

        if(onSocket !=null){
            try {
                onSocket.close();
                onSocket = null;
            }catch (Exception e){
                e.printStackTrace();
            }
        }
        //RaceResultsDisplayService.this.stop();
        // Post to UI that connection is off
        reconnect();
    }

    public Runnable reconnect = new Runnable() {
        @Override
        public void run() {
            connectToDevice(mMacAddress);
        }
    };

    public void reconnect(){
        stopSelf();
//        Log.d(TAG, "Reconnecting in 5 seconds...");
//        mHandler2.postDelayed(reconnect, 1000 * 60 * 2);
    }

    private final static Object obj = new Object();

    public static void write(byte[] out) {
        // Create temporary object
        ConnectedThread r;
        // Synchronize a copy of the ConnectedThread
        synchronized (obj) {
            if (mState != STATE_CONNECTED)
                return;
            r = mConnectedThread;
        }
        // Perform the write unsynchronized
        r.write(out);
    }

    private synchronized void connected(BluetoothSocket mmSocket, BluetoothDevice mmDevice) {
        Log.d(TAG, "Connected");
        // Cancel the thread that completed the connection
        if (mConnectThread != null) {
            mConnectThread.cancel();
            mConnectThread = null;
        }

        // Cancel any thread currently running a connection
        if (mConnectedThread != null) {
            mConnectedThread.cancel();
            mConnectedThread = null;
        }

        mConnectedThread = new ConnectedThread(mmSocket);
        mConnectedThread.start();
        onSocket = mmSocket;
        setState(STATE_CONNECTED);

        // Post to UI that connection is on!


    }

    private class ConnectThread extends Thread {
        private  BluetoothSocket mmSocket;
        private  BluetoothDevice mmDevice;

        public ConnectThread(BluetoothDevice device) {
            this.mmDevice = device;
        }

        @Override
        public void run() {
            setName("ConnectThread");
            mBluetoothAdapter.cancelDiscovery();
            try {
                mmSocket = device.createRfcommSocketToServiceRecord(UUID_SPP);
                mmSocket.connect();
            } catch (IOException e) {
                e.printStackTrace();
                try {
                    mmSocket.close();
                } catch (IOException e1) {
                    e1.printStackTrace();
                }
                connectionFailed();
                return;

            }
            synchronized (PrinterService.this) {
                mConnectThread = null;
            }
            connected(mmSocket, mmDevice);
        }

        public void cancel() {
            try {
                mmSocket.close();
            } catch (IOException e) {
                Log.e(TAG, "close() of connect socket failed", e);
            }
        }

        public void sendData(String data){
            mConnectedThread.write(data.getBytes());
        }
    }

    private class ConnectedThread extends Thread {
        //private final BluetoothSocket mmSocket;
        private final OutputStream mmOutStream;
        private final InputStream mmInStream;

        byte[] buffer = new byte[256];
        int bufferLength;

        private long last_time = 0;
        private boolean ping_sent = false;

        public ConnectedThread(BluetoothSocket socket) {
            //mmSocket = socket;
            OutputStream tmpOut = null;
            InputStream tmpIn = null;
            try {
                tmpOut = socket.getOutputStream();
                tmpIn = socket.getInputStream();
            } catch (IOException e) {
                Log.e(TAG, "temp sockets not created", e);
            }
            mmOutStream = tmpOut;
            mmInStream = tmpIn;
        }

        @Override
        public void run() {
            Log.i(TAG, "BEGIN mConnectedThread");
            byte[] buffer = new byte[1024];
            int bytes;

            // Keep listening to the InputStream while connected
            while (true) {
                try {
                    // Read from the InputStream
                    bytes = mmInStream.read(buffer);
                    // Send the obtained bytes to the UI Activity
                    mHandler.obtainMessage(msgTypes.MESSAGE_READ, bytes, -1, buffer).sendToTarget();
                } catch (IOException e) {
                    Log.e(TAG, "disconnected", e);
                    connectionLost();
                    break;
                }
            }
        }
        /*public void run() {

            while (mState==STATE_CONNECTED) {


                long time = System.nanoTime();
                if (time-last_time > 10000000000L){
                    if (ping_sent) {
                        Log.d(TAG, "PING NOT RETURNED");
                        connectionLost();
                    } else {
                        last_time = time;
                        String ping = String.format("{\"type\":\"ping\",\"time\":%d}", time);
                        Log.d(TAG, ping);
                        write(ping.getBytes());
                        ping_sent = true;
                    }
                }

                try {
                    if (mmInStream.available()>0){
                        bufferLength = mmInStream.read(buffer);

                        byte[] data = new byte[bufferLength];
                        System.arraycopy(buffer, 0, data, 0, bufferLength);
                        final String response = new String(data, "UTF-8");
                        Log.d(TAG, "R:"+response);
                        System.out.println("R:"+response);
                        if (response.equals(String.format("%d", last_time))) {
                            ping_sent = false;
                        }
                    }

                } catch (IOException e){
                    e.printStackTrace();
                }
            }
        }*/


        public void write(byte[] buffer) {
            try {
                mmOutStream.write(buffer);
            } catch (IOException e) {
                Log.e(TAG, "Exception during write", e);
            }
        }

        public void cancel() {
            try {
                //mmSocket.close();
                setState(STATE_NONE);

            } catch (Exception e) {
                e.printStackTrace();
            }
        }

    }

    @Override
    public void onDestroy() {
        this.unregisterReceiver(onBroadcast);


        setState(STATE_NONE);

        stop();
        super.onDestroy();
    }

    // Binding for Service->UI Communication
    private BroadcastReceiver onBroadcast = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {

            if (intent.hasExtra(BROADCAST_CLASS+"TO_PRINT")){
                if (mState != STATE_CONNECTED) return;

                Bundle extras = intent.getExtras();
                System.out.println("is here "+ extras.toString());
                String data = extras.getString(BROADCAST_CLASS+"TO_PRINT");
                if (data == null){
                    return;
                }

                if (data.equals("true")){
                    try {
                        String bill = extras.getString("bill_to_print");
                        if(bill != null) {
                            mConnectedThread.write(bill.getBytes());
                        }
                    }catch (Exception e){
                        e.printStackTrace();
                    }
                }
            }
        }
    };

}

File: app/src/main/java/com/friendlypos/application/bluetooth/PrinterService.kt
package com.friendlysystemgroup.friendlypos.application.bluetooth

import android.app.Service
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothSocket
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Binder
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.util.Log
import com.friendlysystemgroup.friendlypos.principal.activity.MenuPrincipal
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.UUID

/**
 * Servicio para manejar la comunicación con la impresora Bluetooth
 */
class PrinterService : Service() {
    private var mBluetoothAdapter: BluetoothAdapter? = null
    private var mConnectThread: ConnectThread? = null
    private var mMacAddress: String? = null
    private var mContext: Context? = null
    private var onSocket: BluetoothSocket? = null
    
    override fun onCreate() {
        Log.i(TAG, "Service started")
        registerReceiver(onBroadcast, IntentFilter(BROADCAST_CLASS))
        
        mContext = this
        mHandler2 = Handler(Looper.getMainLooper())
        
        super.onCreate()
    }

    override fun onBind(intent: Intent): IBinder {
        Log.d(TAG, "ONBIND")
        return mBinder
    }

    inner class LocalBinder : Binder() {
        val service: PrinterService
            get() = this@PrinterService
    }

    private val mBinder: IBinder = LocalBinder()

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Log.d(TAG, "Onstart Command")
        
        // Verificar que la intent no sea nula
        if (intent == null) {
            Log.e(TAG, "Intent nula, deteniendo servicio")
            stopSelf()
            return START_NOT_STICKY
        }
        
        mBluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
        val chosenDevice = intent.getStringExtra(BT_DEVICE)
        
        if (mBluetoothAdapter != null && !chosenDevice.isNullOrEmpty()) {
            val pairedDevices = mBluetoothAdapter?.bondedDevices ?: emptySet()
            
            if (pairedDevices.isNotEmpty()) {
                for (d in pairedDevices) {
                    if (d.address == chosenDevice) device = d
                }
            }
            
            if (device == null) {
                Log.d(TAG, "No se encontró el dispositivo, deteniendo servicio")
                return START_NOT_STICKY
            }
            
            deviceName = device?.name
            mMacAddress = device?.address
            
            if (!mMacAddress.isNullOrEmpty()) {
                Log.d(TAG, "Conectando a: $deviceName")
                connectToDevice(mMacAddress)
            } else {
                Log.d(TAG, "No hay dirección MAC, deteniendo servicio")
                stopSelf()
                return START_NOT_STICKY
            }
        } else {
            Log.d(TAG, "No hay adaptador Bluetooth o dispositivo elegido")
            stopSelf()
            return START_NOT_STICKY
        }

        return START_STICKY
    }

    @Synchronized
    private fun connectToDevice(macAddress: String?) {
        if (macAddress.isNullOrEmpty() || mBluetoothAdapter == null) {
            Log.e(TAG, "Dirección MAC nula o adaptador Bluetooth no disponible")
            return
        }
        
        Log.d(TAG, "Conectando... ")
        val device = mBluetoothAdapter?.getRemoteDevice(macAddress)
        
        // Cancela el hilo de conexión si está en progreso
        if (mState == STATE_CONNECTING) {
            mConnectThread?.cancel()
            mConnectThread = null
        }

        // Cancela cualquier hilo que esté ejecutando una conexión
        mConnectedThread?.cancel()
        mConnectedThread = null
        
        // Inicia nuevo hilo de conexión
        device?.let {
            mConnectThread = ConnectThread(it)
            mConnectThread?.start()
            setState(STATE_CONNECTING)
        }
    }

    private fun setState(state: Int) {
        mState = state
    }

    @Synchronized
    fun stop() {
        setState(STATE_NONE)
        mHandler2?.removeCallbacks(reconnect)

        mConnectThread?.cancel()
        mConnectThread = null

        mConnectedThread?.cancel()
        mConnectedThread = null
        
        mBluetoothAdapter?.cancelDiscovery()
        
        stopSelf()
    }

    override fun stopService(name: Intent): Boolean {
        setState(STATE_NONE)
        
        mConnectThread?.cancel()
        mConnectThread = null

        mConnectedThread?.cancel()
        mConnectedThread = null
        
        mBluetoothAdapter?.cancelDiscovery()
        
        return super.stopService(name)
    }

    private fun connectionFailed() {
        Log.d(TAG, "Conexión fallida")
        if (mState == STATE_NONE) return
        reconnect()
    }

    fun connectionLost() {
        Log.d(TAG, "Conexión perdida")
        
        mConnectThread?.cancel()
        mConnectThread = null

        mConnectedThread?.cancel()
        mConnectedThread = null

        try {
            onSocket?.close()
            onSocket = null
        } catch (e: Exception) {
            Log.e(TAG, "Error al cerrar socket", e)
        }
        
        reconnect()
    }

    private val reconnect: Runnable = Runnable { 
        connectToDevice(mMacAddress) 
    }

    fun reconnect() {
        stopSelf()
        // Si se desea reconectar automáticamente después de un tiempo:
        // mHandler2?.postDelayed(reconnect, 5000) // 5 segundos
    }

    @Synchronized
    private fun connected(mmSocket: BluetoothSocket?, mmDevice: BluetoothDevice) {
        if (mmSocket == null) {
            Log.e(TAG, "Socket nulo en connected()")
            connectionFailed()
            return
        }
        
        Log.d(TAG, "Conectado")
        
        // Cancela el hilo que completó la conexión
        mConnectThread?.cancel()
        mConnectThread = null

        // Cancela cualquier hilo que esté ejecutando una conexión
        mConnectedThread?.cancel()
        mConnectedThread = null

        // Inicia el hilo para manejar la conexión
        mConnectedThread = ConnectedThread(mmSocket)
        mConnectedThread?.start()
        onSocket = mmSocket
        setState(STATE_CONNECTED)
    }

    private inner class ConnectThread(private val mmDevice: BluetoothDevice) : Thread() {
        private var mmSocket: BluetoothSocket? = null

        override fun run() {
            name = "ConnectThread"
            mBluetoothAdapter?.cancelDiscovery()
            
            try {
                mmSocket = device?.createRfcommSocketToServiceRecord(UUID_SPP)
                mmSocket?.connect()
            } catch (e: IOException) {
                Log.e(TAG, "Error al conectar socket", e)
                try {
                    mmSocket?.close()
                } catch (e1: IOException) {
                    Log.e(TAG, "Error al cerrar socket después de fallo de conexión", e1)
                }
                connectionFailed()
                return
            }
            
            synchronized(this@PrinterService) {
                mConnectThread = null
            }
            
            connected(mmSocket, mmDevice)
        }

        fun cancel() {
            try {
                mmSocket?.close()
            } catch (e: IOException) {
                Log.e(TAG, "Error al cerrar socket de conexión", e)
            }
        }

        fun sendData(data: String) {
            mConnectedThread?.write(data.toByteArray())
        }
    }

    private inner class ConnectedThread(socket: BluetoothSocket) : Thread() {
        private val mmOutStream: OutputStream?
        private val mmInStream: InputStream?

        init {
            var tmpOut: OutputStream? = null
            var tmpIn: InputStream? = null
            
            try {
                tmpOut = socket.outputStream
                tmpIn = socket.inputStream
            } catch (e: IOException) {
                Log.e(TAG, "Error al crear streams temporales", e)
            }
            
            mmOutStream = tmpOut
            mmInStream = tmpIn
        }

        override fun run() {
            Log.i(TAG, "Iniciando hilo conectado")
            val buffer = ByteArray(1024)
            var bytes: Int

            // Escuchar el InputStream mientras esté conectado
            while (true) {
                try {
                    // Leer del InputStream
                    mmInStream?.let {
                        bytes = it.read(buffer)
                        // Enviar los bytes obtenidos a la actividad UI
                        mHandler?.obtainMessage(MESSAGE_READ, bytes, -1, buffer)
                            ?.sendToTarget()
                    } ?: break
                } catch (e: IOException) {
                    Log.e(TAG, "Desconectado", e)
                    connectionLost()
                    break
                }
            }
        }

        fun write(buffer: ByteArray?) {
            if (buffer == null) {
                Log.e(TAG, "Intento de escritura con buffer nulo")
                return
            }
            
            try {
                mmOutStream?.write(buffer)
            } catch (e: IOException) {
                Log.e(TAG, "Excepción durante la escritura", e)
            }
        }

        fun cancel() {
            try {
                setState(STATE_NONE)
            } catch (e: Exception) {
                Log.e(TAG, "Error al cancelar ConnectedThread", e)
            }
        }
    }

    override fun onDestroy() {
        try {
            unregisterReceiver(onBroadcast)
        } catch (e: Exception) {
            Log.e(TAG, "Error al desregistrar receptor", e)
        }

        setState(STATE_NONE)
        stop()
        super.onDestroy()
    }

    // Receptor para la comunicación Servicio->UI
    private val onBroadcast: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (intent.hasExtra("$BROADCAST_CLASS$TO_PRINT")) {
                if (mState != STATE_CONNECTED) {
                    Log.d(TAG, "No conectado, ignorando mensaje para imprimir")
                    return
                }

                val extras = intent.extras ?: return
                val data = extras.getString("$BROADCAST_CLASS$TO_PRINT") ?: return

                if (data == "true") {
                    try {
                        val bill = extras.getString("bill_to_print")
                        if (!bill.isNullOrEmpty()) {
                            mConnectedThread?.write(bill.toByteArray())
                        } else {
                            Log.d(TAG, "Factura para imprimir está vacía")
                        }
                    } catch (e: Exception) {
                        Log.e(TAG, "Error al procesar factura para imprimir", e)
                    }
                }
            }
        }
    }

    companion object {
        const val CLASS_NAME = "com.friendlypos.application.bluetooth.PrinterService"
        const val BROADCAST_CLASS = "com.friendlypos.printbill"
        private const val TO_PRINT = "TO_PRINT"
        private val UUID_SPP = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB")
        
        // Constantes de estado
        const val STATE_NONE = 0 // No haciendo nada
        const val STATE_LISTEN = 1 // Escuchando conexiones entrantes
        const val STATE_CONNECTING = 2 // Iniciando una conexión saliente
        const val STATE_CONNECTED = 3 // Conectado a un dispositivo remoto
        
        // Tipos de mensajes
        const val MESSAGE_READ = 1
        
        // Constante para el dispositivo Bluetooth
        const val BT_DEVICE = "btdevice"

        private const val TAG = "PrinterService"
        
        // Variables compartidas
        private var mConnectedThread: ConnectedThread? = null
        var mHandler: Handler? = null
        var mHandler2: Handler? = null
        var mState = STATE_NONE
        var deviceName: String? = null
        var device: BluetoothDevice? = null
        private val syncObject = Any()

        /**
         * Detiene el servicio de impresora Bluetooth
         */
        fun stop(context: MenuPrincipal): Boolean {
            if (context.isServiceRunning(CLASS_NAME)) {
                val serviceIntent = Intent(context, PrinterService::class.java)
                context.stopService(serviceIntent)
                return true
            }
            return false
        }

        /**
         * Inicia el servicio de impresora Bluetooth
         */
        fun startRDService(context: Context, prefExternalDisplay: String?) {
            if (prefExternalDisplay.isNullOrEmpty()) {
                Log.d(TAG, "No hay dirección de impresora configurada")
                return
            }

            val serviceIntent = Intent(context, PrinterService::class.java).apply {
                putExtra(BT_DEVICE, prefExternalDisplay)
            }
            context.startService(serviceIntent)
        }

        /**
         * Escribe datos en la impresora conectada
         */
        fun write(out: ByteArray?) {
            if (out == null) {
                Log.e(TAG, "Intento de escritura con datos nulos")
                return
            }
            
            // Crea un objeto temporal
            val r: ConnectedThread?
            
            // Sincroniza una copia del ConnectedThread
            synchronized(syncObject) {
                if (mState != STATE_CONNECTED) {
                    Log.d(TAG, "No conectado, no se puede escribir")
                    return
                }
                r = mConnectedThread
            }
            
            // Realiza la escritura no sincronizada
            r?.write(out)
        }
    }
}